src/gf.c

```diff
@@ -2233,7 +2233,30 @@ static jl_tupletype_t *lookup_arg_type_tuple(jl_value_t *arg1 JL_PROPAGATES_ROOT
     return jl_lookup_arg_tuple_type(arg1, args, nargs, 1);
 }
 
-jl_method_instance_t *jl_method_lookup(jl_value_t **args, size_t nargs, size_t world)
+JL_DLLEXPORT jl_method_instance_t *jl_method_lookup_by_tt(jl_tupletype_t *tt, size_t world, jl_value_t *_mt)
+{
+    jl_methtable_t *mt = NULL;
+    if (_mt == jl_nothing)
+        mt = jl_gf_ft_mtable(jl_tparam0(tt));
+    else {
+        assert(jl_isa(_mt, jl_methtable_type));
+        mt = (jl_methtable_t*) _mt;
+    }
+
+    // TODO: The fast path is heavily optimized for args and avoids forming the signature type
+    jl_genericmemory_t *leafcache = jl_atomic_load_relaxed(&mt->leafcache);
+    jl_typemap_entry_t *entry = lookup_leafcache(leafcache, (jl_value_t*)tt, world);
+    if (entry)
+        return entry->func.linfo;
+    JL_GC_PUSH1(&tt);
+    JL_LOCK(&mt->writelock);
+    jl_method_instance_t *sf = jl_mt_assoc_by_type(mt, tt, world);
+    JL_UNLOCK(&mt->writelock);
+    JL_GC_POP();
+    return sf;
```

Comment:
It looks like we should move the lock acquire inside this method, since this is a lot of code for all 3 callers to duplicate each time
```suggestion
    return jl_mt_assoc_by_type(mt, tt, world);
```